homeassistant/config.py

```diff
@@ -1015,21 +1029,115 @@ async def merge_packages_config(
+        line: int | str = "?"
+        if (log_message := p_ex.log_message) is None:
+            # No pre defined log_message is set, so we generate a message
+            if isinstance(ex, vol.Invalid):
+                log_message = format_schema_error(hass, ex, p_name, p_config, link)
+                if annotation := find_annotation(p_config, ex.path):
+                    config_file, line = annotation
+            elif isinstance(ex, HomeAssistantError):
+                log_message = format_homeassistant_error(
+                    hass, ex, p_name, p_config, link
+                )
+            else:
+                log_message = str(ex)
+            config_error_messages.append((domain, p_ex, log_message, config_file, line))
+        else:
+            general_error_messages.append((domain, p_ex))
+        _LOGGER.error(log_message, exc_info=p_ex.ex if p_ex.log_exception else None)
+
+    if not raise_on_failure or not config_exception_info:
+        return
+
+    placeholders: dict[str, str]
+    if len(config_error_messages) == 1 and not general_error_messages:
+        domain, p_ex, log_message, config_file, line = config_error_messages[0]
+        ex = p_ex.ex
+        p_name = p_ex.platform_name
+        translation_key = p_ex.translation_key
+        placeholders = {
+            "domain": domain,
+            "p_name": p_name,
+            "error": str(ex),
+            "errors": str(len(config_exception_info)),
+            "config_file": config_file,
+            "line": str(line),
+        }
+    elif len(general_error_messages) == 1 and not config_error_messages:
+        domain, p_ex = general_error_messages[0]
+        ex = p_ex.ex
+        translation_key = p_ex.translation_key
+        log_message = str(p_ex.log_message)
+        placeholders = {
+            "domain": domain,
+            "error": str(ex),
+            "errors": str(len(config_exception_info)),
+        }
+    else:
+        translation_key = "integration_config_error"
+        errors = str(len(config_exception_info))
+        log_message = (
+            f"Failed to process component config for integration {integration.domain} "
+            f"due to multiple errors ({errors}), check the logs for more information."
+        )
+        placeholders = {
+            "domain": integration.domain,
+            "errors": errors,
+        }
+    raise ConfigValidationError(
+        log_message,
+        [p_ex.ex for p_ex in config_exception_info],
+        translation_domain="homeassistant",
+        translation_key=translation_key,
+        translation_placeholders=placeholders,
+    )
+
+
 async def async_process_component_config(  # noqa: C901
-    hass: HomeAssistant, config: ConfigType, integration: Integration
-) -> ConfigType | None:
-    """Check component configuration and return processed configuration.
+    hass: HomeAssistant,
+    config: ConfigType,
+    integration: Integration,
+) -> tuple[ConfigType | None, list[ConfigExceptionInfo]]:
```

Comment:
If we want to return something more here in the future, we're going to break it again. Should we return a dataclass instead?

Reply:
If we make `async_handle_component_config_errors` to wrap `async_process_component_config` then this would make sense to the second element in the tuple. The first element allows direct assigning the config.